packages/flutter/lib/src/material/material.dart

```diff
@@ -617,27 +628,37 @@ abstract class InkFeature {
   /// Called when the ink feature is no longer visible on the material.
   final VoidCallback? onRemoved;
 
-  bool _debugDisposed = false;
+  bool _disposed = false;
+
+  /// Whether or not visual reaction is activated.
+  ///
+  /// Change this field will affect whether this InkFeature is render in next
+  /// frame.
+  ///
+  /// For this InkFeature to render properly, it should usually be change in
+  /// [State.deactivate] and [State.activate].
+  bool visible = true;
 
   /// Free up the resources associated with this ink feature.
+  ///
+  /// This method can only be called once per [InkFeature] instance.
   @mustCallSuper
   void dispose() {
-    assert(!_debugDisposed);
-    assert(() {
-      _debugDisposed = true;
-      return true;
-    }());
+    assert(!_disposed);
+    _disposed = true;
     _controller._removeFeature(this);
     onRemoved?.call();
   }
 
   void _paint(Canvas canvas) {
     assert(referenceBox.attached);
-    assert(!_debugDisposed);
+    assert(!_disposed);
     // find the chain of renderers from us to the feature's referenceBox
     final List<RenderObject> descendants = <RenderObject>[referenceBox];
     RenderObject node = referenceBox;
     while (node != _controller) {
+      if (node.parent == null)
+        return;
```

Comment:
Isn't it a bug if we were told to paint, but we can't find the controller in our ancestor change? Why is it ok to silently not paint?

Reply:
Good point. I'm still figuring out why this change causes crash on [this minimized test case](https://github.com/flutter/flutter/pull/81445/files#diff-04e0901809bec10776d61f4d7d4d681fe0c101e445ca8ccee9fb9ceca6c551fcR1442).